### ISSUE TITLE
Make citus_prepare_pg_upgrade fails on non-existent pg_dist_object entries

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#5069]
We observed an issue during citus_prepare_pg_upgrade where pg_dist_object references a non-existent objects:

```sql                                                                                          
UPDATE citus.pg_dist_object SET (type, object_names, object_args) = (SELECT * FROM pg_identify_object_as_address(classid, objid, objsubid))
ERROR:  invalid role OID: 73072   
```

This may have been due to a combination of CREATE/DROP ROLE and setting changes.

Related issue: we observed workers have an arbitrary set of pg_dist_object records.

**Comment by ozgune** [comment]
Hi @marcocitus - thank you for opening this issue. @metdos , all, do we have a sense of how long it would take to fix the issue?

**Comment by metdos** [comment]
> Hi @marcocitus - thank you for opening this issue. @metdos , all, do we have a sense of how long it would take to fix the issue?

Our expectation is around 1 week.

**Comment by ivyazmitinov** [comment]
Hi!
We faced a similar issue with `citus_activate_node` on 11.1.4:
```sql
SELECT * FROM citus_activate_node('<node_id>', 5432);
# ERROR:  invalid role OID: 1003588
```

I assume, a manual fix of the `pg_dist_object` should work?

**Comment by thanodnl** [comment]
Removing the non-existing roles from `pg_dist_object` should allow the node to be added. It is important to make sure to _only_ remove that/those row(s).

Citus uses this table to keep track of all objects it should synchronize with the workers when they are altered and as a source of which objects to create when adding new nodes. Removing too many entries might have negative effects in scaling out as well.

Do you have any idea how it could be that when the role was removed it wasn't removed from `pg_dist_node`? Any chance you have an extra database that you connect to when adding/removing users?

**Comment by ivyazmitinov** [comment]
@thanodnl 

> Any chance you have an extra database that you connect to when adding/removing users?

Yep, that's exactly what happened: we created users on the `postgres` database and later recreated on another

**Comment by thanodnl** [comment]
Creation doesn't make much difference. Once you assign permissions for that user to read/write distributed tables Citus will recognise the user has been created in an other database and creates the user on the worker nodes just in time.

However when that user gets deleted from a separate database there is no way for Citus to understand this user has been deleted, hence the kept reference in pg_dist_object.

We currently advise to only connect to one database when using Citus for this reason. Performing management operations to the postgres 'cluster' (that is the very unfortunate name postgres uses to describe how to run multiple databases under one postmaster) from a different database could cause such issues in Citus. Extensions are only active inside a databases and thus Citus cannot (today) work well with multiple databases.

**Comment by ivyazmitinov** [comment]
> Extensions are only active inside a databases and thus Citus cannot (today) work well with multiple databases.

Wish we knew that before splitting data into 30+ databases :upside_down_face:
